src/psyclone/psyir/transformations/debug_checksum_trans.py

```diff
@@ -0,0 +1,203 @@
+
+'''This module contains the DebugChecksumTrans class.'''
+
+from typing import Union, List
+
+from psyclone.psyir.frontend.fortran import FortranReader
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.psyir.nodes import Assignment, Node, Reference, Routine
+from psyclone.psyir.nodes.structure_accessor_mixin import (
+    StructureAccessorMixin
+)
+from psyclone.psyir.symbols import (
+    DataSymbol, INTEGER_TYPE, PreprocessorInterface, ScalarType
+)
+from psyclone.psyir.transformations.region_trans import RegionTrans
+
+
+class DebugChecksumTrans(RegionTrans):
+    '''
+    Creates a set of checksums (written via print) for all written arrays
+    inside the provided region.
+
+    For example:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> from psyclone.transformations import DebugChecksumTrans
+
+    >>> psyir = FortranReader().psyir_from_source("""
+    ...     subroutine mysubroutine()
+    ...     integer, dimension(10,10) :: A
+    ...     integer :: i
+    ...     integer :: j
+    ...     do i = 1, 10
+    ...       do j = 1, 10
+    ...         A(i,j) = A(i,k) + i-j
+    ...       end do
+    ...     end do
+    ...     end subroutine
+    ...     """)
+    ... loop = psyir.children[0].children[0]
+    ... DebugChecksumTrans().apply(loop)
+    ... print(FortranWriter()(psyir))
+    subroutine mysubroutine()
+      integer, dimension(10,10) :: a
+      integer :: i
+      integer :: j
+      integer :: PSYCLONE_INTERNAL_line_
+    <BLANKLINE>
+      do i = 1, 10, 1
+        do j = 1, 10, 1
+          a(i,j) = a(i,j) + i - j
+        enddo
+      enddo
+      PSYCLONE_INTERNAL_line_ = __LINE__
+      PRINT *, "PSyclone checksums from mysubroutine at line:", \
+PSYCLONE_INTERNAL_line_ + 1
+      PRINT *, "a checksum", SUM(a)
+    <BLANKLINE>
+    end subroutine mysubroutine
+    <BLANKLINE>
+
+    '''
+    def apply(self, node: Union[Node, List[Node]], options=None) -> None:
+        '''
+        Applies the checksum transformation to the provided node(s).
+
+        :param node: The node or list of nodes to apply the
+                      transformation to.
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node, options={"node-type-check": False})
+
+        node_list = self.get_node_list(node)
+        routine = node_list[0].ancestor(Routine)
+        routine_table = routine.symbol_table
+
+        fwriter = FortranWriter()
+        writes = []
+        # Loop over the assignments in the region
+        assigns = []
+        for node in node_list:
+            assigns.extend(node.walk(Assignment))
+        # Loop through the assignments and find the arrays
+        for assign in assigns:
+            if isinstance(assign.lhs, StructureAccessorMixin):
+                # Find the last member.
```

Comment:
You shouldn't have to do this - `assign.lhs.datatype` *should* be all that you need? If it isn't, please could you explain what the problem is?

Reply:
Added comments to why this is done.

Reply:
Great, thanks.